src/katgpucbf/xbgpu/engine.py

```diff
@@ -259,9 +273,171 @@ def capture_enable(self, enable: bool = True) -> None:
+            # TODO: Clarify shape of `saturated`
+            saturated = accel.DeviceArray(
+                context, shape=(engine.heaps_per_fengine_per_chunk, self.n_beams), dtype=np.uint32
+            )
+            present_ants = np.zeros(shape=(engine.n_ants,), dtype=bool)
+            tx_item = TxQueueItem(buffer_device, saturated, present_ants)
+            self._tx_free_item_queue.put_nowait(tx_item)
+
+        # TODO: The way this is imported will change once the OperationSequence is in place
+        self.send_stream = bsend.BSend(
+            output,
+            engine.heaps_per_fengine_per_chunk,
+            self.n_tx_items,
+            n_channels_per_substream=engine.n_channels_per_substream,
+            spectra_per_heap=engine.src_layout.n_spectra_per_heap,
+            timestamp_step=engine.rx_heap_timestamp_step,
+            send_rate_factor=engine.send_rate_factor,
+            channel_offset=engine.channel_offset_value,
+            context=context,
+            packet_payload=engine.dst_packet_payload,
+            stream_factory=lambda stream_config, buffers: make_bstream(
+                endpoints=[output.dst],
+                interface=engine.dst_interface,
+                ttl=engine.dst_ttl,
+                use_ibv=engine.dst_ibv,
+                affinity=engine.dst_affinity,
+                comp_vector=engine.dst_comp_vector,
+                stream_config=stream_config,
+                buffers=buffers,
+            ),
+            tx_enabled=init_tx_enabled,
+        )
+
+        self._populate_sensors()
+
+    def _populate_sensors(self) -> None:
+        sensors = self.engine.sensors
+        # Dynamic sensors
+        # TODO: Update once multiple beams are supported
+        sensors.add(
+            aiokatcp.Sensor(
+                int,
+                f"{self.output.name}-beng-clip-cnt",
+                "Number of complex samples that saturated.",
+                default=0,
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        )
+
+    def capture_enable(self, enable: bool = True) -> None:  # noqa: D102
+        # NOTE: Defaulting to handle its single output for now.
+        # Actual implementation will require some kind of 'stream_id' arg,
+        # which currently contravenes the base function signature.
+        self.send_stream.enable_substream(enable=enable)
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
+        while True:
+            # Get item from the receiver loop.
+            # - Wait for the HtoD transfers to complete, then
+            # - Give the chunk back to the receiver for reuse.
+            rx_item = await self._rx_item_queue.get()
+            if rx_item is None:
+                break
+            await rx_item.async_wait_for_events()
+
+            tx_item = await self._tx_free_item_queue.get()
+            await tx_item.async_wait_for_events()
```

Comment:
Can just do
```suggestion
            tx_item.enqueue_wait_for_events(self._proc_command_queue)
```
That ensures the GPU waits, without holding up the Python code.

Reply:
`enqueue_wait_for_events` only blocks the GPU stream. It returns immediately on the CPU, unlike `await async_wait_for_events` which prevents the next line of code from running until the work is complete.

Reply:
OK, I had misunderstood the distinction there.
The way Amish has it now is the way it's done in the XPipeline, so if there's an issue it may be worth addressing there as well. (In a separate ticket of course.)

Reply:
Yes I think I've noticed that code in XPipeline and it's bugged me. It's not wrong, and it probably makes no difference, but it's not quite as efficient as it could be.

Reply:
I'd suggest then leaving this one as-is for the time being, and filing a technical debt ticket to improve both this one and the Xpipeline one.

I don't like it when a PR goes through lots of rounds of review with many many changes requested, it makes keeping track of individual issues quite difficult.

Reply:
Fair points all around (I genuinely thought I'd addressed this in a previous commit-review by Bruce).

I've created [NGC-1145](https://skaafrica.atlassian.net/browse/NGC-1145) in the meantime.